File: frontend/src/types/person_document.ts
/**
 * Kişisel evrak tip tanımları — backend PersonDocumentOut ve
 * HealthDocumentSummaryOut şemalarıyla senkron.
 */

export type PersonDocumentType =
  | 'profile_photo'
  | 'identity_copy'
  | 'health_report'
  | 'parental_permission'
  | 'undertaking'
  | 'waiver'
  | 'other'

export type ReviewStatus =
  | 'pending'
  | 'approved'
  | 'rejected'
  | 'expired'
  | 'superseded'

export type ScanStatus =
  | 'pending'
  | 'clean'
  | 'infected'
  | 'failed'
  | 'skipped_dev'

export type DeleteRequestStatus = 'pending' | 'approved' | 'rejected'

export interface DeleteRequest {
  reason: string
  requested_by_user_id: string
  status: DeleteRequestStatus
  created_at: string // ISO8601 datetime
}

export interface PersonDocumentOut {
  id: string
  club_id: string
  subject_person_id: string
  guardian_link_id: string | null
  document_type: PersonDocumentType
  original_filename: string
  mime_type: string
  size_bytes: number
  uploaded_at: string // ISO8601 datetime
  valid_until: string | null // ISO8601 date
  retain_until: string | null // ISO8601 date
  review_status: ReviewStatus
  scan_status: ScanStatus
  is_sensitive: boolean
  is_deleted: boolean
  supersedes_id: string | null
  rejection_reason: string | null
  reviewed_at: string | null // ISO8601 datetime
  processing_basis: string | null
  delete_request: DeleteRequest | null
}

export interface HealthDocumentSummaryOut {
  subject_person_id: string
  exists: boolean
  valid_until: string | null // ISO8601 date
}

export interface DeleteRequestOut {
  id: string
  document_id: string
  requested_by_user_id: string
  reason: string
  created_at: string // ISO8601 datetime
  status: DeleteRequestStatus
}

File: backend/app/schemas/person_document.py
"""Kişisel evrak API şemaları — storage ayrıntıları özellikle dışarı verilmez."""
import json
import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, field_validator


class DeleteRequestData(BaseModel):
    """Belge üzerindeki silme isteği — DB'de JSON metin olarak saklanır."""
    model_config = ConfigDict(from_attributes=True)

    reason: str = ""
    requested_by_user_id: uuid.UUID
    status: str  # "pending" | "approved" | "rejected"
    created_at: datetime


class PersonDocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID
    subject_person_id: uuid.UUID
    guardian_link_id: Optional[uuid.UUID] = None
    document_type: str
    original_filename: str
    mime_type: str
    size_bytes: int
    uploaded_at: datetime
    valid_until: Optional[date] = None
    retain_until: Optional[date] = None
    review_status: str
    scan_status: str
    is_sensitive: bool
    is_deleted: bool = False
    supersedes_id: Optional[uuid.UUID] = None
    rejection_reason: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    processing_basis: Optional[str] = None
    delete_request: Optional[DeleteRequestData] = None

    @field_validator("delete_request", mode="before")
    @classmethod
    def _parse_delete_request_json(cls, v: object) -> object:
        """DB'den gelen JSON metin string'ini dict'e çevirir."""
        if isinstance(v, str):
            return json.loads(v)
        return v


class RejectionBody(BaseModel):
    rejection_reason: str


class HealthDocumentSummaryOut(BaseModel):
    subject_person_id: uuid.UUID
    exists: bool
    valid_until: Optional[date] = None


class DeleteRequestBody(BaseModel):
    reason: str = ""


class DeleteRequestOut(BaseModel):
    id: uuid.UUID
    document_id: uuid.UUID
    requested_by_user_id: uuid.UUID
    reason: str
    created_at: datetime
    status: str  # "pending" | "approved" | "rejected"
